### ISSUE TITLE
`permute_vertices` is confusing: it uses the inverse permutation

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1930]
**tl;dr**

 - `permute_vertices` uses the inverse of the permutation it is given to permute vertices.
 - The documentation of `permute_vertices` is incorrect, in all interfaces.
 - The way `permute_vertices` works is confusing, inconvenient, and inconsistent with other parts of the library.

----

### The problem

Regarding `igraph_permute_vertices()`, in [the Python docs](https://igraph.org/python/api/latest/igraph._igraph.GraphBase.html#permute_vertices) it says:

> Permutes the vertices of the graph according to the given permutation and returns the new graph.

In R,

> permutation | A numeric vector giving the permutation to apply.

**But that's not what it does. It actually applies the inverse permutation.**

In Mathematica, I use a separate implementation instead of `igraph_permute_vertices()`, so I won't discuss that.

**Note:** In this issue, I will use 1-based indexing.

Consider this graph. I marked both names and indices on the vertices:

<img width="360" alt="image" src="https://user-images.githubusercontent.com/1212871/149617630-1740ef93-5fc6-42dd-b08d-6cf64ad44a69.png">

If we were to apply the permutation `(2, 3, 1)`, we would get:

<img width="360" alt="image" src="https://user-images.githubusercontent.com/1212871/149617677-aa13709e-c48f-4fb5-b242-e5e56ef2e3c3.png">

In a high-level interface, this could be expressed as either of these:

 - Permutation to apply to vertices: Apply the permutation `2, 3, 1`. These integers can also be interpreted as the vertex indices in the original graph.
 - New name order: I want the new vertex order to be `b, c, a`.

The original order is `a,b,c`. Applying the permutation `2,3,1` gives the new order `b,c,a`. The inverse permutation of `2,3,1` is `3,1,2`. Applying this to the new order returns the original order.

However, what `permute_vertices()` does is _not_ apply a permutation, as I showed here. Instead, if applies the inverse. This is not what the R and Python docs say. What about the C docs? They are somewhat more accurate, but also a bit confused. They say:

> permutation The permutation to apply.

This is very misleading. But then it goes on:

> Vertex 0 is mapped to the first element of the vector, vertex 1 to the second, etc. 

So, if the input is `3,1,2` (note: 1-based), then the "mapping" is as

```
name  old_index  new_index
a        1         3
b        2         1
c        3         2
```

`3,1,2` is the inverse permutation from the first example above. It achieves the new vertex order `b,c,a`. These numbers can be understood as vertex indices _in the new graph_, not in the old.

### Why is the direct (not inverse) permutation is better

I think it would be better if this function took the direct permutation as input.

 - First, this is probably more intuitive for most people, or at least a much better match for the name of the function.

 - This allows for more flexibility in high-level interfaces. Vertex indices _in the new graph_ (i.e. inverse permutation) are not easy to use before the new graph exists. It is not possible to specify them in terms of "vertex selectors" before the new graph exists. It is not possible to specify them in terms of vertex names at all. 

    If we use vertex indices in the old graph (i.e. direct permutation), then they can easily be given in terms of _either_ vertex names _or_ indices.

 - I would expect to be able to use the output of `igraph_canonical_permutation()` in `igraph_permute_vertices()`, but this is not possible. We need to invert it first.

 - The attribute handler interface has a function to permute attributes (see `igraph_i_attribute_permute_vertices()` and the `.permute_vertices()` member of the attribute table). It uses direct permutations. `igraph_permute_vertices()` is not consistent with this. In fact, `igraph_permute_vertices()` needs to internally invert the permutation so that it can call `igraph_i_attribute_permute_vertices()`.

 - See [this example](https://stackoverflow.com/a/70720310/695132) for how one would generally end up with the direct permutation in many use cases.

### Proposal

I wish I could propose to simply change how `igraph_permute_vertices()` behaves, but this is very dangerous. Even if we are fine with breaking API, keeping a function's name and signature, but changing what it does, is very likely to lead to lead to bugs in someone's code (likely also in igraph internals).

Instead, I would propose deprecating `igraph_permute_vertices()` (because its name is confusing), and introducing two new functions: one that works the same as `igraph_permute_vertices()` and one that uses the direct permutation. Proposals for the names are welcome. I don't have anything I am 100% happy with, but one possibility for the inverse permutation version is `igraph_remap_vertex_indices()` while for the direct permutation we can use `igraph_reorder_vertices()`.

**Comment by ntamas** [comment]
Well, personally I've always found permutations confusing because both representations (the direct and the inverse) seem equally logical to me when thinking strictly about mappings from {1..n} to {1..n}. One particular representation was chosen for `igraph_permute_vertices()`, and I was always thinking about this representation as "the" permutation because it is equivalent to the "one-line notation" as outlined in [Wikipedia](https://www.wikiwand.com/en/Permutation#/One-line_notation). So it wasn't just an arbitrary choice, although I find it surprising and unfortunate that the result of `igraph_canonical_permutation()` is _not_ appropriate as an input for `igraph_permute_vertices()`.

So, the question is: is there a consensus in either math or CS about how a permutation should be represented? For me it does not seem like there is any (in CS I have seen both, and in math it seems like the cycle representation is the most popular, but this is probably less intuitive to end users and is harder to work with algorithmically), so what we should aim for at best is _internal_ consistency within igraph.

**Comment by szhorvat** [comment]
As far as I can tell, what's shown in your Wikipedia link is not what `permute_vertices()` uses, but what I proposed it _should_ be using. Am I getting it wrong?

In programming languages where it is possible to do indexing using a vector of indices (Mathematica or R), the most convenient representation is the one that allows for permuting an array using `arr[permutation]`. This is not what `permute_vertices()` does.  Python is more verbose, but it should be possible to use `[array[i] for i in permutation]`.

**Comment by ntamas** [comment]
> As far as I can tell, what's shown in your Wikipedia link is not what permute_vertices() uses, but what I proposed it should be using. Am I getting it wrong?

Here's an example from Python:

```
>>> from igraph import Graph
>>> g = Graph(4)
>>> g.vs["name"] = ["a", "b", "c", "d"]
>>> g2 = g.permute_vertices([1, 2, 3, 0])
```

Here, the permutation vector means that the "image" of vertex 0 after the permutation is vertex 1, the "image" of vertex 1 is vertex 2, the "image" of vertex 2 is vertex 3 and the "image" of vertex 3 is vertex 0. So, the expected vector of `g2.vs["name"]` should be `["d", "a", "b", "c"]`, and this is what I get:

```
>>> g2.vs["name"]
['d', 'a', 'b', 'c']
```

This is consistent with Wikipedia in my opinion. Item `i` in the permutation vector describes the "image" of `i`, i.e. the result of `\sigma(i)` where `\sigma` is the mapping that takes an item from the "domain" set (i.e. an integer from 0..|V|-1 that refers to the original graph) and it produces an item in the "image" set (which is also an integer from 0..|V|-1).

**Comment by szhorvat** [comment]
We need to make a decision on this. @vtraag, can you take a look and give an opinion?

**Comment by szhorvat** [comment]
**This is work-in-progress, please do not comment until done!**

### Introduction

Permutations are very commonly used in graph theory. We should use a consistent representation for them in igraph, and a consistent terminology to talk about them. If a user received a permutation from another library, or wants to pass a permutation generated by igraph to another library, it should be clear how to convert them. Each igraph function that takes a permutation or produces a permutation should use the same representation. I hope we agree on this.

### Definitions

First, what is a permutation, and what does one do with it?

Let's phrase it like this: a permutation $p$ is a reordering operation that can be applied to a list/vector $x$ to obtain the list $y$.

```
permute(x, p)  --->  y
```

There are multiple ways to represent `p`, including cycle notation, but for most purposes it is simplest to use a list of integer indices. I will use 1-based indices in the examples below. Then we have two representation choices:

 - **Representation 1** (R1): The elements of `p` are indices into `x`. The `i`th element of `y` is taken from position `p[i]` in `x`. The action of the permutation is conveniently computed by iterating through `y` and taking elements from `x` as `y[i] = x[ p[i] ]`.

 - **Representation 2** (R2): The elements of `p` are indices into `y`. The `i`th element of `x` is sent to position `p[i]` in `y`. The action of the permutation is conveniently computed by iterating through `x` and sending element to `y` as `y[ p[i] ] = x[i]`.

Note that if an index vector `p` can be interpreted as the permutation $P$ according to R1 then it represents the inverse $P^{-1}$ according to R2, and vice versa.

### Potential terminology

In the context of the input vector `x`, we may refer to the entries of `p` as:

 - R1: _Indices_, as they are indices into `x`. In the specific context of vertex lists, these are vertex IDs.
 - R2: _Ranks_, as in `y` lowest rank comes first and highest rank last.

### Comparing representations

Is R1 or R2 more convenient from a programming perspective? This probably differs from application to application. When using _an explicit loop_ to compute the action of permutations, they are equally convenient.

However, many high-level languages provide an array indexing operation. In such languages, R1 has the advantage that one can simply write `y = x[p]`. However, one could always write an explicit function for performing reordering, `y = permute(x, p)`.

One nice property of R1 that R2 does not possess is that in the context of named vertex lists, the permutation could be specified in terms of vertex *names* instead of *indices*. Suppose the vertex list is `(a, b, c)`. Then, if we want to obtain the vertex order `(c, a, b)`, a hypothetical vertex permutation function could be called _either_ as `permute_vertices(graph, (c, a, b))` or as `permute_vertices(graph, (3, 1, 2))`.  With the R2 representation, using vertex _names_ is not meaningful as the indices in `p` are not IDs in the output graph, not the input graph.

The R1 representation has the convenience that it can be generalized for subsetting lists with incomplete index sets, for example: `(a, b, c)[(3, 2)] --> (c, b)`. This is not possible with R2. Of course, `(3, 2)` is not a proper permutation.

The R1 representation has the convenience that it remains interpretable when the `p` vector contains duplicate indices. For example, `(a, b, c)[(3, 1, 3)] --> (c, a, c)`. While this is again not a proper permutation, one may argue that it is a meaningful operation for a ...

TODO: In-place permutation algorithm?

TODO: Inversion algorithm?

TODO: In-place inversion algorithm?

### What representations do other systems use?

 - Mathematica uses R2 (besides cycle notation), and provides [`Permute[]`](http://reference.wolfram.com/language/ref/Permute.html).  Thus `Permute[x, p]` and array-indexing `x[[p]]` do not produce the same result. However, Mathematica also has several functions which effectively produce R1 representations to be used in an `x[[p]]` manner, for example, `x[[ Ordering[x] ]]` sorts `x`. `Ordering` is stated to return a "list of positions" though, not a "permutation".
 - Maple TODO
 - SageMath TODO
 - GAP TODO
 - Magma TODO
 - Bliss TODO (because we interface with Bliss)


### What igraph functions operate with permutations?

Which representation does each existing igraph function use?

 - `igraph_permute_vertices()` uses **R2**.
 - `igraph_automorphism_group()` TODO.
 - `igraph_vector_order()` produces the permutation that sorts a vector. There are several similar functions. Documentation seems to state R2, need to verify.
 - `igraph_maximum_cardinality_search()` produces an R1 representation in `alpham1` and an R2 representation in `alpha`.
 - `igraph_vector_permute()` **R1**
 - `igraph_vector_ptr_permute()` TODO, probably R1 as well?
 - `igraph_vector_list_permute()` TODO, probably R1 as well?
 - `igraph_sparsemat_permute()` TODO, seems R1, need to check carefully. The underlying `cs_permute` function uses the inverse permutation for row-indices and direct for column indices!
 - attribute handler `permute_vertices()` TODO
 - attribute handler `permute_edges()` TODO

**Comment by ntamas** [comment]
Also, note that there's `igraph_vector_index()`, which uses a sort-of-kind-of a generalized R1 permutation. And this way it's consistent with `igraph_vector_permute()`; the difference is that `igraph_vector_permute()` is in-place, while `igraph_vector_index()` creates a new vector.

**Comment by ntamas** [comment]
One other nice side effect of changing `igraph_permute_vertices()` to return an R1 permutation would be to allow the output of `igraph_topological_sorting()` to be fed directly into `igraph_permute_vertices()` to obtain a graph that is already sorted topologically.

**Comment by ntamas** [comment]
@szhorvat I am trying to bring the `develop` branch of the Python interface up-to-date with the `develop` branch of the C core and it seems like the output of `igraph_canonical_permutation()` _was_ consistent with `igraph_permute_vertices()`. At least I think so because unit tests related to the canonical permutation broke with this change. From the documentation of `igraph_canonical_permutation()` it also seems that it is returning a permutation that is _now_ incompatible with `igraph_permute_vertices()`.

Can you please double-check because now I am getting a bit confused.

Also, `igraph_invert_permutation()` is now exposed in 4301f6e76 so if we see that this needs to be fixed by changing the output of `igraph_canonical_permutation()` it is now easier to do so.

**Comment by ntamas** [comment]
Okay, the problem is (probably) different. But I fail to understand why this unit test fails now in Python:

```python
        g = Graph.GRG(10, 0.5)
        perm = list(range(10))
        shuffle(perm)
        g2 = g.permute_vertices(perm)
        g3 = g.permute_vertices(g.canonical_permutation())
        g4 = g2.permute_vertices(g2.canonical_permutation())

        self.assertTrue(g3.vcount() == g4.vcount())
        self.assertTrue(sorted(g3.get_edgelist()) == sorted(g4.get_edgelist()))
```

I checked that `g` and `g2` are isomorphic (as they should be because I am simply permuting the vertices of `g` to obtain `g2`). If they are isomorphic, then applying their canonical permutation to them should yield identical graphs, right? Yet they do not seem to be identical - the last assertion fails.

**Comment by szhorvat** [comment]
As I remember the canonical permutation function inverted internally? I am not able to check the code now, unfortunately. 

I'll have a look next week. 

We should definitely provide a helper function that inverts and validates permutations.

**Comment by ntamas** [comment]
All sorted out now. It turned out that when changing the semantics of the permutation in `permute_vertices()` we forgot to update `canonical_permutation()` to do the same. Now it's okay.
